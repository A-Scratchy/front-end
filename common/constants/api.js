import { get, post, patch } from 'common/utils/api-utils';
import { coerceEmptyStringToUndefined } from 'common/utils/string-utils';

/* GET REQUESTS */
export const getCodeSchoolsPromise = () => get('code_schools');
export const getMentorPromise = id => get(`mentors/${id}`);
export const getMentorsPromise = () => get('mentors');
export const getRequestsPromise = () => get('requests');
export const getScholarshipPromise = id => get(`scholarships/${id}`);
export const getScholarshipsPromise = () => get('scholarships');
export const getServicesPromise = () => get('services');
export const getTeamMembersPromise = () => get('team_members');

/* POST REQUESTS */
export const createUser = ({ email, password, firstName, lastName, zipcode }) =>
  post('users', {
    user: {
      email,
      password,
      first_name: firstName,
      last_name: lastName,
      zip: zipcode,
    },
  }).then(({ data }) => data);

export const loginUser = ({ email, password }) =>
  post('sessions', {
    user: {
      email,
      password,
    },
  }).then(({ data }) => data);

export const postMentorRequestPromise = ({ language, additionalDetails, mentor, service }) =>
  post('requests', {
    request: {
      details: additionalDetails,
      language,
      requested_mentor_id: mentor,
      service_id: service,
    },
  });

/* PATCH REQUESTS */
export const updateUser = ({
  branchOfService,
  companyName,
  companyRole,
  disciplines,
  doesWantMentor, // TODO: Use in form
  doesWantScholarshipInfo, // TODO: Use in form
  doesWantToVolunteer, // TODO: Use in form
  employmentStatus,
  militaryStatus,
  payGrade,
  programmingLanguages,
  yearsOfService,
}) => {
  const interestItems = [...programmingLanguages, ...disciplines];

  // Interests could be an array of many filled strings, one filled string, or one empty string
  const interests =
    interestItems.length > 1
      ? interestItems.join(', ')
      : [coerceEmptyStringToUndefined(interestItems[0])];

  return patch('users', {
    user: {
      branch_of_service: coerceEmptyStringToUndefined(branchOfService),
      company_name: coerceEmptyStringToUndefined(companyName),
      company_role: coerceEmptyStringToUndefined(companyRole),
      mentor: doesWantMentor,
      scholarship_info: doesWantScholarshipInfo,
      volunteer: doesWantToVolunteer,
      employment_status: coerceEmptyStringToUndefined(employmentStatus),
      military_status:
        militaryStatus === 'civilian' ? undefined : coerceEmptyStringToUndefined(militaryStatus),
      pay_grade: coerceEmptyStringToUndefined(payGrade),
      interests,
      years_of_service: coerceEmptyStringToUndefined(yearsOfService),
    },
  }).then(({ data }) => data);
};

export const patchUpdateMentorRequestPromise = ({ request, status, mentor }) =>
  patch(`requests/${request}`, {
    request: {
      status,
      mentor,
    },
  });
