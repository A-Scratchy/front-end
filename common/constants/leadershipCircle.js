// Instructions:

// Below is a list. To do this, you must do the following:
// 1:
// 2:
// 3:
// 4:
// 5:
// 6:
// 7:

const honorRoll = [{}];

export default honorRoll;
