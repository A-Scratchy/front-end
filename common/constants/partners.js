const partners = [
  {
    name: 'APEX Systems',
    logo: 'partnerLogos/apex_systems.png',
    url: 'https://apexsystems.com',
  },
  {
    name: 'GitHub',
    logo: 'partnerLogos/github.png',
    url: 'https://github.com',
  },
  {
    name: '1Password',
    logo: 'partnerLogos/1password.png',
    url: 'https://1password.com/',
  },
  {
    name: 'Docker',
    logo: 'partnerLogos/docker.png',
    url: 'https://docker.com',
  },
  {
    name: 'Oracle',
    logo: 'partnerLogos/oracle.png',
    url: 'https://oracle.com',
  },
  {
    name: "O'Reilly Media",
    logo: 'partnerLogos/oreilly.png',
    url: 'https://oreilly.com',
  },
  {
    name: 'HackerRank',
    logo: 'partnerLogos/hackerrank.png',
    url: 'https://hackerrank.com',
  },
  {
    name: 'Zapier',
    logo: 'partnerLogos/zapier.png',
    url: 'https://zapier.com',
  },
];

export default partners;
