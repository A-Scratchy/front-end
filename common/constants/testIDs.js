export const ALERT = 'ALERT';
export const ALERT_CLOSE_BUTTON = 'Close Alert Button';
export const BUTTON = 'BUTTON';
export const CHANGE_PASSWORD_FORM_ERROR = 'CHANGE_PASSWORD_FORM_ERROR';
export const FLAT_CARD_IMAGE = 'FLAT_CARD_IMAGE';
export const INPUT = 'INPUT';
export const INPUT_ERROR = 'INPUT_ERROR';
export const INPUT_FEEDBACK_GROUPING = 'INPUT_FEEDBACK_GROUPING';
export const LABEL = 'LABEL';
export const SCREEN_READER_ONLY = 'SCREEN_READER_ONLY';
