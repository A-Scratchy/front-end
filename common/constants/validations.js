export const minPasswordCharNum = 8;

export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailMatch: 'Emails must match',
  password: 'Must include the following: lowercase letter, uppercase letter, number',
  passwordMatch: 'Passwords must match',
  length: val => `Must be at least ${val} characters`,
  zipcode: 'Must be a valid zipcode',
};
