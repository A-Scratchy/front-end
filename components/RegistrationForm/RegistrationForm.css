.RegistrationForm {
  width: 100%;
}

.row {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}
