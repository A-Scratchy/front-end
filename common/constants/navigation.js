import { donateLink } from 'common/constants/urls';

// links shared between nav and footer (no duplicate code)
const contactLink = {
  name: 'Contact Us',
  href: '/contact',
};

const faqLink = {
  name: 'FAQ',
  href: '/faq',
};

const getInvolvedLink = {
  name: 'Get Involved',
  href: '/get_involved',
};

const eventsLink = {
  name: 'Events',
  href: '/events',
};

export const whoWeServeSection = {
  name: 'Who We Serve',
  href: '/who_we_serve',
  shouldPrefetch: false,
  sublinks: [
    {
      name: 'Member Login',
      href: '/login',
    },
    {
      name: 'Join',
      href: '/join',
    },
  ],
};

const logoutLink = {
  name: 'Logout',
  href: '/login?loggedOut=true',
  shouldPrefetch: false,
  sublinks: [],
};

export const navItems = [
  {
    name: 'About Us',
    href: '/about',
    shouldPrefetch: false,
    sublinks: [contactLink, faqLink],
  },
  {
    ...eventsLink,
    shouldPrefetch: false,
    sublinks: [],
  },
  {
    ...getInvolvedLink,
    shouldPrefetch: false,
    sublinks: [
      // {
      //   name: 'Mentoring',
      //   href: '/mentoring',
      // },
      {
        name: 'Sponsorship',
        href: '/sponsorship',
      },
      {
        name: 'Leadership Circle',
        href: '/leadership_circle',
      },
    ],
  },
];

export const loggedInNavItems = [...navItems, logoutLink];
export const loggedOutNavItems = [...navItems.slice(0, 1), whoWeServeSection, ...navItems.slice(1)];

export const footerItems = {
  column1: [
    {
      href: '/about',
      name: 'About Us',
    },
    contactLink,
    faqLink,
  ],
  column2: [
    {
      href: '/code_schools',
      name: 'Code Schools',
    },
    {
      href: '/jobs',
      name: 'Job Board',
    },
    eventsLink,
    {
      href: '/who_we_serve',
      name: 'Who We Serve',
    },
  ],
  column3: [
    getInvolvedLink,
    // {
    //   href: '/mentoring',
    //   name: 'Become A Mentor',
    // },
    {
      href: '/history',
      name: 'History',
    },
    {
      href: donateLink,
      name: 'Donate',
      analyticsEventLabel: 'Donate',
    },
  ],
  column4: [
    // {
    //   href: '/resources',
    //   name: 'Resources',
    // },
    {
      href: '/press',
      name: 'Press',
    },
    {
      href: '/branding',
      name: 'Branding',
    },
    {
      href: '/team',
      name: 'Team',
    },
    // {
    //   href: '/blog',
    //   name: 'Blog',
    // },
  ],
  legal: [
    {
      href: '/terms',
      name: 'Terms of Use',
    },
    // TODO: Create a link for this
    // {
    //   href: 'https://www.anotherfakeurl.com/',
    //   name: 'Cookies',
    //   analyticsEventLabel: 'Cookies',
    // },
    {
      // NOTE: If you change this route, please update the redirect in `now.json` as well
      href: 'https://www.iubenda.com/privacy-policy/8174861',
      name: 'Privacy',
      analyticsEventLabel: 'Privacy',
    },
  ],
};
