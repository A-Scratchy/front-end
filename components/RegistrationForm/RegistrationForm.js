import React, { Component } from 'react';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import { minPasswordCharNum, validationErrorMessages } from 'common/constants/validations';
import { isMinPasswordStrength } from 'common/utils/validator-utils';
import Button from 'components/_common_/Button/Button';
import Form from 'components/_common_/Form/Form';
import Input from 'components/_common_/Form/Input/Input';

const registrationSchema = Yup.object().shape({
  email: Yup.string()
    .required('Required')
    .email(validationErrorMessages.email),
  'confirm-email': Yup.string()
    .required('Required')
    .oneOf([Yup.ref('email')], validationErrorMessages.emailMatch),
  password: Yup.string()
    .required('Required')
    .min(minPasswordCharNum, validationErrorMessages.length(minPasswordCharNum))
    .test('password-strength', validationErrorMessages.password, isMinPasswordStrength),
  'confirm-password': Yup.string()
    .required('Required')
    .oneOf([Yup.ref('password')], validationErrorMessages.passwordMatch),
});

// TODO: Define links for terms of service / privacy policy and place on this page

export default class RegistrationForm extends Component {
  state = {
    // firstName: '',
    // lastName: '',
    email: '',
    // zip: '',
    password: '',
    // isMentor: '',
  };

  render() {
    const { state } = this;

    return (
      <Formik
        initialValues={{ ...state }}
        onSubmit={(values, { setSubmitting }) => {
          // Probably HTTP request right here...
          setTimeout(() => {
            console.log(JSON.stringify(values, null, 2));
            setSubmitting(false);
          }, 400);
        }}
        validationSchema={registrationSchema}
      >
        {({ isSubmitting }) => (
          <Form>
            <Field
              type="email"
              name="email"
              label="Email"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="email"
              name="confirm-email"
              label="Confirm Email"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="password"
              name="password"
              label="Password"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              type="password"
              name="confirm-password"
              label="Confirm Password"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Button type="submit" disabled={isSubmitting}>
              Submit
            </Button>
          </Form>
        )}
      </Formik>
    );
  }
}
