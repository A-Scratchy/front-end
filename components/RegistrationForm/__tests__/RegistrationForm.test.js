import React from 'react';
import { mount } from 'enzyme';
import { createUser } from 'common/constants/api';
import { validationErrorMessages } from 'common/constants/validations';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import wait from 'test-utils/wait';
import RegistrationForm from '../RegistrationForm';

const asyncRenderDiff = async enzymeWrapper => {
  await wait();
  enzymeWrapper.update();
};

afterEach(() => {
  OperationCodeAPIMock.reset();
});

describe('RegistrationForm', () => {
  it('should render with required props', () => {
    createSnapshotTest(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);
  });

  it('should display required error message when blurring past email input', async () => {
    const wrapper = mount(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);

    wrapper.find('input#email').simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.required);
  });

  it('should show error when providing non-email to email input', async () => {
    const wrapper = mount(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);

    wrapper
      .find('input#email')
      .simulate('change', { target: { id: 'email', value: 'email' } })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.email);
  });

  it('should show "password required" message when blurring past input', async () => {
    const wrapper = mount(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);

    wrapper.find('input#password').simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="password"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.required);
  });

  it('should show "invalid password" message when focusing off an invalid password', async () => {
    const wrapper = mount(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);

    const stringWithoutNumber = 'SillyPassword';

    wrapper
      .find('input#password')
      .simulate('change', { target: { id: 'password', value: stringWithoutNumber } })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="password"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.password);
  });

  it('should display password match message when both password inputs do not match', async () => {
    const wrapper = mount(<RegistrationForm register={jest.fn()} onSuccess={jest.fn()} />);

    wrapper
      .find('input#password')
      .simulate('change', {
        target: { id: 'password', value: 'ValidPassword1' },
      })
      .simulate('blur');

    wrapper
      .find('input#confirm-password')
      .simulate('change', {
        target: { id: 'confirm-password', value: 'something' },
      })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('input#confirm-password')
        .closest('Input')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.passwordMatch);
  });

  it('should submit with valid data in form', async () => {
    const successSpy = jest.fn();
    const wrapper = mount(
      <RegistrationForm
        onSuccess={successSpy}
        register={jest.fn()}
        initialValues={{
          email: 'email@email.com',
          'confirm-email': 'email@email.com',
          password: 'abc123ABC',
          'confirm-password': 'abc123ABC',
          firstName: 'Test',
          lastName: 'User',
          zipcode: 90630,
        }}
      />,
    );

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(successSpy).toHaveBeenCalled();
  });

  it('should NOT submit with invalid data in form', async () => {
    const successSpy = jest.fn();

    const wrapper = mount(
      <RegistrationForm
        onSuccess={successSpy}
        register={jest.fn()}
        initialValues={{
          email: 'email@email.com',
          'confirm-email': 'ffdsdsfsadf@fdsafdsafsd.com',
          password: 'abc1231111',
          'confirm-password': '111111',
          firstName: '',
          lastName: '',
          zipcode: '',
        }}
      />,
    );

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(successSpy).not.toHaveBeenCalled();
  });

  it('should show "email already registered" message for dupe email registration', async () => {
    const user = {
      email: 'kylemh.email12@gmail.com',
      password: 'Testing123!',
      firstName: 'Test',
      lastName: 'User',
      zipcode: 90630,
    };

    OperationCodeAPIMock.onAny().reply(422, { email: ['has been taken'] });

    const successSpy = jest.fn();
    const wrapper = mount(
      <RegistrationForm
        register={createUser}
        onSuccess={successSpy}
        initialValues={{
          ...user,
          'confirm-email': user.email,
          'confirm-password': user.password,
        }}
      />,
    );

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(successSpy).not.toHaveBeenCalled();
    expect(
      wrapper
        .find('Alert')
        .last()
        .text(),
    ).toStrictEqual('Email has been taken.');
  });
});
