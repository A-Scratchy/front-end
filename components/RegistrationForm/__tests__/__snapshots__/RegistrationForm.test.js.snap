// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm should render with required props 1`] = `
<form
  className="RegistrationForm"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="email"
      >
        Email*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="username email"
          className="input"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="confirm-email"
      >
        Confirm Email*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="username email"
          className="input"
          disabled={false}
          id="confirm-email"
          name="confirm-email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="password"
      >
        Password*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="new-password"
          className="input"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="confirm-password"
      >
        Confirm Password*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="new-password"
          className="input"
          disabled={false}
          id="confirm-password"
          name="confirm-password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="firstName"
      >
        First Name*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="given-name"
          className="input"
          disabled={false}
          id="firstName"
          name="firstName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="lastName"
      >
        Last Name*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="family-name"
          className="input"
          disabled={false}
          id="lastName"
          name="lastName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="zipcode"
      >
        Zipcode*
      </label>
      <div
        className="inputFeedBackGrouping"
      >
        <input
          autoComplete="postal-code"
          className="input"
          disabled={false}
          id="zipcode"
          name="zipcode"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="Alert error"
      role="alert"
    >
      
    </div>
  </div>
  <div
    className="row"
  >
    <button
      className="Button secondary"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
